client/consensus/manual-seal/src/heartbeat_stream.rs

```diff
@@ -0,0 +1,133 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
+const DEFAULT_MAX_BLOCKTIME: u64 = 30;
+const DEFAULT_MIN_BLOCKTIME: u64 = 1;
+const DEFAULT_FINALIZE: bool = false;
+
+/// Heartbeat options to manage the behavior of the heartbeat stream
+pub struct HeartbeatOptions {
+	/// The minimum amount of time to pass before generating a new block, in sec.
+	pub min_blocktime: u64,
+	/// The amount of time passed that a new heartbeat block will be generated, in sec.
+	pub max_blocktime: u64,
+	/// Control whether the generated block is finalized
+	pub finalize: bool,
+}
+
+impl Default for HeartbeatOptions {
+	fn default() -> Self {
+		Self {
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			max_blocktime: DEFAULT_MAX_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+/// HeartbeatStream is combining transaction pool notification stream and generate a new block
+///   when a certain time has passed without any transactions.
+pub struct HeartbeatStream<Hash> {
+	/// The transaction pool notification stream
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	/// Future to control when the next block should be generated
+	delay_future: Delay,
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
+}
+
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		opts: HeartbeatOptions
+	) -> Self {
+		if opts.min_blocktime > opts.max_blocktime {
+			panic!("Heartbeat options `min_blocktime` value must not be larger than `max_blocktime` value.");
+		}
+		Self {
+			pool_stream,
+			delay_future: tokio::time::delay_for(Duration::from_secs(opts.max_blocktime)),
+			last_blocktime: None,
+			opts
+		}
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		let mut hbs = self.get_mut();
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				if let Some(last_blocktime) = hbs.last_blocktime {
+					// If the last block happened less than the min_blocktime time, we want to wait till
+					//   `min_blocktime` has lapsed.
+					let passed_blocktime = Instant::now() - last_blocktime;
+					if passed_blocktime < Duration::from_secs(hbs.opts.min_blocktime) {
+						// We set `delay_future` here so it will wake up when the min_blocktime since the last
+						//   block created is passed.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime)
+							- passed_blocktime);
+						return Poll::Pending;
```

Comment:
This will not work. When you create a new future, you need to poll it, so it is registered to wake up the future.

Reply:
The latest commit should fix this issue.